File: algorithms/implementation/breaking-best-and-worst-records/breaking-best-and-worst-records-python/tests/breakingRecords.py

# O problema "Breaking the Records" 
# em HackerRank solicita a implementação 
# de uma função chamada breakingRecords(scores), 
# onde scores é um array representando as 
# pontuações obtidas em uma série de jogos. 
# O objetivo é contar quantas vezes o jogador 
# quebra seu próprio recorde de pontuação 
# mais alta e mais baixa durante os jogos.

# def breakingRecords(pontuacoes):
    
#     # Inicializa as variáveis para armazenar as pontuações máxima e mínima.
#     max_pontucao = pontuacoes[0]
#     min_pontuacao = pontuacoes[0]
    
#     # Inicializa as variáveis para contar as quebras de recordes.
#     max_conta = 0
#     min_conta = 0
    
#     # Itera sobre as pontuações para verificar os recordes.
#     for pontuacao in pontuacoes:
        
#         # Verifica se a pontuação atual é maior que a pontuação máxima registrada.
#         if pontuacao > max_pontucao:
            
#             # Atualiza a pontuação máxima e incrementa o contador de quebras de recorde máximo.
#             max_pontucao = pontuacao
#             max_conta += 1
        
#         # Verifica se a pontuação atual é menor que a pontuação mínima registrada.
#         elif pontuacao < min_pontuacao:
            
#             # Atualiza a pontuação mínima e incrementa o contador de quebras de recorde mínimo.
#             min_pontuacao = pontuacao
            
#             min_conta += 1
    
#     # Retorna um array contendo o número de quebras de recorde máximo e mínimo, respectivamente.
#     return [max_conta, min_conta]


def breakingRecords(pontuacoes):
    
    """
    Função que conta quantas vezes os recordes máximo e mínimo foram quebrados.
    
    Parameters:
    pontuacoes (List[int]): Lista de pontuações.
    
    Returns:
    List[int]: Lista com a contagem de quebras do recorde máximo e mínimo.
    """
    
    # Inicializa as variáveis para armazenar a maior e menor pontuação, 
    # e os contadores de quebra de recorde.
    maior_pontuacao = [pontuacoes[0], 0]
    menor_pontuacao = [pontuacoes[0], 0]
    
    # Percorre as pontuações para contar as quebras de recordes.
    for i in pontuacoes[1:]:
        
        # Verifica se a pontuação atual é menor que o recorde mínimo.
        if i < menor_pontuacao[0]:
            
            # Atualiza o recorde mínimo e incrementa o contador.
            menor_pontuacao = [i, menor_pontuacao[1] + 1]
            
        # Verifica se a pontuação atual é maior que o recorde máximo.
        if i > maior_pontuacao[0]:
            
            # Atualiza o recorde máximo e incrementa o contador.
            maior_pontuacao = [i, maior_pontuacao[1] + 1]
    
    # Retorna uma lista com a contagem de quebras do recorde máximo e mínimo.
    return [maior_pontuacao[1], menor_pontuacao[1]]

File: algorithms/implementation/breaking-best-and-worst-records/breaking-best-and-worst-records-javascript/js/breakingRecords.js
// # O problema "Breaking the Records" 
// # em HackerRank solicita a implementação 
// # de uma função chamada breakingRecords(scores), 
// # onde scores é um array representando as 
// # pontuações obtidas em uma série de jogos. 
// # O objetivo é contar quantas vezes o jogador 
// # quebra seu próprio recorde de pontuação 
// # mais alta e mais baixa durante os jogos.


// /**
//  * Função que conta quantas vezes os recordes máximo e mínimo foram quebrados.
//  * @param {number[]} pontuacoes - Array de pontuações.
//  * @returns {number[]} - Array com a contagem de quebras do recorde máximo e mínimo.
//  */

// export function breakingRecords(pontuacoes) {

//     // Inicializa as pontuações atuais para o primeiro elemento do array.
//     let atual_alto = pontuacoes[0];
//     let atual_baixo = pontuacoes[0];

//     // Inicializa contadores de quebra de recordes.
//     let contar_alto = 0;
//     let contar_baixo = 0;

//     // Percorre as pontuações para contar as quebras de recordes.
//     pontuacoes.forEach(pontuacao => {

//         // Verifica se a pontuação atual é maior que o recorde máximo.
//         if (pontuacao > atual_alto) {
            
//             // Atualiza o recorde máximo.
//             atual_alto = pontuacao;

//             // Incrementa o contador de quebra de recorde máximo.
//             contar_alto += 1;

//         } else if (pontuacao < atual_baixo) {
            
//             // Atualiza o recorde mínimo.
//             atual_baixo = pontuacao;

//             // Incrementa o contador de quebra de recorde mínimo.
//             contar_baixo += 1;
//         }
//     });

//     // Retorna um array com a contagem de quebras do recorde máximo e mínimo.
//     return [contar_alto, contar_baixo];
// }


// /**
//  * Função que conta quantas vezes os recordes máximo e mínimo foram quebrados.
//  *
//  * @param {number[]} pontuacoes - Lista de pontuações.
//  * @returns {number[]} - Lista com a contagem de quebras do recorde máximo e mínimo.
//  */
// export function breakingRecords(pontuacoes) {

//     // Inicializa um array para armazenar a contagem de quebras do recorde máximo e mínimo.
//     const arr = [0, 0]

//     // Inicializa as variáveis para armazenar a menor e maior pontuação.
//     let min = pontuacoes[0];
//     let max = pontuacoes[0];

//     // Percorre as pontuações para contar as quebras de recordes.
//     for (const pontuacao of pontuacoes) {
        
//         // Verifica se a pontuação atual é menor que o recorde mínimo.
//         if (pontuacao < min) {
            
//             // Atualiza o recorde mínimo.
//             min = pontuacao;

//             // Incrementa o contador de quebra do recorde mínimo.
//             arr[1] += 1;
//         }

//         // Verifica se a pontuação atual é maior que o recorde máximo.
//         if (pontuacao > max) {
            
//             // Atualiza o recorde máximo.
//             max = pontuacao

//             // Incrementa o contador de quebra do recorde máximo.
//             arr[0] += 1;
//         }
//     }

//     // Retorna o array com a contagem de quebras do recorde máximo e mínimo.
//     return arr
// }


/**
 * Função que conta quantas vezes os recordes máximo e mínimo foram quebrados.
 *
 * @param {number[]} pontuacoes - Lista de pontuações.
 * @returns {number[]} - Lista com a contagem de quebras do recorde máximo e mínimo.
 */

export function breakingRecords(pontuacoes) {

    // Utiliza a função reduce para iterar sobre as pontuações e acumular as quebras de recordes.
    return pontuacoes.reduce((recorde, pontuacao) => {

        // Verifica se a pontuação atual é maior que o recorde máximo.
        if(pontuacao > recorde[0]) {

            // Atualiza o recorde máximo.
            recorde[0] = pontuacao;

            // Incrementa o contador de quebra do recorde máximo.
            recorde[2] += 1;
        }

        // Verifica se a pontuação atual é menor que o recorde mínimo.
        if (pontuacao < recorde[1]) {
            
            // Atualiza o recorde mínimo.
            recorde[1] = pontuacao;

            // Incrementa o contador de quebra do recorde mínimo.
            recorde[3] += 1;
        }

        // Retorna o acumulador.
        return recorde;

        // Inicializa o acumulador com valores iniciais e extrai o resultado final.
    }, [pontuacoes[0], pontuacoes[0], 0, 0]).splice(2);
}